runtime/shared_common/CacheMap.cpp

```diff
@@ -444,148 +475,236 @@ SH_CacheMap::startup(J9VMThread* currentThread, J9SharedClassPreinitConfig* pico
-			 */
-			if (sanityWalkROMClassSegment(currentThread, _ccHead) == 0) {
-				rc = CC_STARTUP_CORRUPT;
-			}
-		}
-		if (rc == CC_STARTUP_OK) {
-			/* THREADING: We want the cache mutex here as we are reading all available data. Don't want updates happening as we read. */
-			if (_ccHead->enterWriteMutex(currentThread, false, fnName) == 0) {
-				 /* populate the hashtables */
-				itemsRead = readCache(currentThread, _ccHead, -1, false);
+		IDATA tryCntr = 0;
+		bool isCcHead = (ccToUse == _ccHead);
+		const char* cacheUniqueIDPtr = NULL;
 
-				_ccHead->protectPartiallyFilledPages(currentThread);
+		/* start up _ccHead (the top layer cache) and then statrt its pre-requiste cache (ccNext). Contine to startup ccNext and its pre-requiste cache, util there is no more pre-requiste cache.
+		 *     _ccHead -------------> ccNext ---------> ccNext --------> ........---------> ccTail
+		 *   (top layer)          (middle layer)     (middle layer)      ........         (layer 0)
+		 */
 
-				if ((*_runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_REDUCE_STORE_CONTENTION) && !_ccHead->isRunningReadOnly()) {
-					_ccHead->setWriteHash(currentThread, 0);				/* Initialize to zero so that peek will work */
-				}
+		if (!isCcHead) {
+			runtimeFlags = &_sharedClassConfig->readOnlyCacheRuntimeFlags;
+		}
 
-				/* Two reasons for moving the code to check for full cache from SH_CompositeCacheImpl::startup()
-				 * to SH_CacheMap::startup():
-				 * 	- While marking cache full, last unused pages are also protected, which ideally should be done
-				 * 	  after protecting pages belonging to ROMClass area and metadata area.
-				 * 	- Secondly, when setting cache full flags, the code expects to be holding the write mutex, which is not done in
-				 * 	  SH_CompositeCacheImpl::startup().
-				 *
-				 * Do not call fillCacheIfNearlyFull() in readonly mode, as we cannot write anything to cache.
-				 */
-				if ((0 == (*_runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_NESTED))
-					&& (!_ccHead->isRunningReadOnly())
-				) {
-					_ccHead->fillCacheIfNearlyFull(currentThread);
-				}
+		do {
+			++tryCntr;
+			if ((rc == CC_STARTUP_SOFT_RESET) && (deleteRC == -1)) {
+				/* If we've tried SOFT_RESET the first time around and the delete failed,
+			 	 * remove AUTOKILL so that we start up with the existing cache */
+				*runtimeFlags &= ~J9SHR_RUNTIMEFLAG_AUTOKILL_DIFF_BUILDID;
+			}
 
-				_ccHead->exitWriteMutex(currentThread, fnName);
+			rc = ccToUse->startup(currentThread, piconfig, cacheMemoryUT, runtimeFlags, _verboseFlags, _cacheName, cacheDirName, cacheDirPerm, &_actualSize, &_localCrashCntr, true, cacheHasIntegrity);
+			if (rc == CC_STARTUP_OK) {
+				if (sanityWalkROMClassSegment(currentThread, ccToUse) == 0) {
+					rc = CC_STARTUP_CORRUPT;
+					goto error;
+				}
 
-				if (CM_READ_CACHE_FAILED == itemsRead) {
-					Trc_SHR_CM_startup_Exit6(currentThread);
-					return -1;
+				if (isCcHead) {
+					if (!initializeROMSegmentList(currentThread)) {
```

Comment:
This seems too early, all the caches need to be validated before the ROM segments are added. If something goes wrong, the JVM can still run if nonFatal is used.

Reply:
When creating a new layer, getPrereqCache() will add the PrereqCacheID to the top layer cache. ROM segments should be already initialized before we call into getPrereqCache(), so I put initializeROMSegmentList() here.

Reply:
I modified the layer 0 cache and started up the the layer 1 and 0 cache again with nonfatal, it works fine:

```
java -Xshareclasses:name=CC1,nonfatal -version
JVMSHRC838E Failed to start up the shared cache. Modification to a lower layer shared cache has been detected. Expected Unique Cache ID is /tmp/javasharedresources/C290M4F1A64P_CC1_G39L00-1000000_5d38c980
JVMSHRC686I Failed to startup shared class cache. Continue without using it as -Xshareclasses:nonfatal is specified
java version "1.8.0_221"
...
...
...
```

Reply:
It might appear to work, but it isn't correct to have segments describing classes that can never be used. At the best is it confusing for service, at the worst if the cache is unmapped from memory the ROM segments could overlap with other memory allocations and cause runtime problems.

Reply:
Moved initializeROMSegmentList() to the end of startup().